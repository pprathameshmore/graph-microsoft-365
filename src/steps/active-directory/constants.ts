import {
  RelationshipClass,
  StepEntityMetadata,
  StepRelationshipMetadata,
} from '@jupiterone/integration-sdk-core';

export const steps: Record<string, string> = {
  FETCH_ACCOUNT: 'account',
  FETCH_GROUPS: 'groups',
  FETCH_GROUP_MEMBERS: 'group-members',
  FETCH_USERS: 'users',
};

export const entities: Record<string, StepEntityMetadata> = {
  ACCOUNT: {
    resourceName: '[AD] Account',
    _type: 'microsoft_365_account',
    _class: 'Account',
  },
  GROUP: {
    resourceName: '[AD] Group',
    _type: 'azure_user_group',
    _class: 'UserGroup',
  },
  USER: {
    resourceName: '[AD] User',
    _type: 'azure_user',
    _class: 'User',
  },
  ORGANIZATION: {
    resourceName: '[AD] Organization',
    _type: 'azure_organization',
    _class: 'Organization',
  },
  /**
   * The entity used for members of groups which are not one of the ingested
   * directory objects.
   */
  GROUP_MEMEBER: {
    resourceName: '[AD] Group Member',
    _type: 'azure_group_member',
    _class: 'User',
  },
};

export const relationships: Record<string, StepRelationshipMetadata> = {
  ACCOUNT_HAS_USER: {
    _type: 'microsoft_365_account_has_azure_user',
    sourceType: entities.ACCOUNT._type,
    _class: RelationshipClass.HAS,
    targetType: entities.USER._type,
  },
  ACCOUNT_HAS_GROUP: {
    _type: 'microsoft_365_account_has_azure_group',
    sourceType: entities.ACCOUNT._type,
    _class: RelationshipClass.HAS,
    targetType: entities.GROUP._type,
  },
  GROUP_HAS_MEMBER: {
    _type: 'azure_group_has_member',
    sourceType: entities.GROUP._type,
    _class: RelationshipClass.HAS,
    targetType: entities.GROUP_MEMEBER._type,
  },
  GROUP_HAS_USER: {
    _type: 'azure_group_has_user',
    sourceType: entities.GROUP._type,
    _class: RelationshipClass.HAS,
    targetType: entities.USER._type,
  },
  GROUP_HAS_GROUP: {
    _type: 'azure_group_has_group',
    sourceType: entities.GROUP._type,
    _class: RelationshipClass.HAS,
    targetType: entities.GROUP._type,
  },
};

export const DATA_ACCOUNT_ENTITY = entities.ACCOUNT._type;
