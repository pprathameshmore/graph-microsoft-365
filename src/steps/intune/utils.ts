import { IntegrationLogger } from '@jupiterone/integration-sdk-core';
import { DeviceConfigurationDeviceStatus } from '@microsoft/microsoft-graph-types-beta';

/**
 * Device statuses that indicate that a finding should not be open
 */
export const CLOSED_DEVICE_STATUSES: DeviceConfigurationDeviceStatus['status'][] = [
  'notApplicable',
  'compliant',
  'remediated',
  'notAssigned',
];

/**
 * Device statuses that indicate that a finding should not be open
 */
export const OPEN_DEVICE_STATUSES: DeviceConfigurationDeviceStatus['status'][] = [
  'nonCompliant',
  'error',
  'conflict',
  'unknown',
];

/**
 * Device statuses that indicate that there is not a relationship between a device and config
 */
export const UNRELATED_DEVICE_STATUSES: DeviceConfigurationDeviceStatus['status'][] = [
  'notAssigned',
  'notApplicable',
];

/**
 * Determines if a device is related to a configuration based on the device state status on the config
 */
export function deviceIsRelatedToConfig(
  status: DeviceConfigurationDeviceStatus['status'],
) {
  return ![...UNRELATED_DEVICE_STATUSES, undefined].includes(status);
}

/**
 * Determines if the finding is open based of the device status
 */
export function findingIsOpen(
  deviceStatus: DeviceConfigurationDeviceStatus['status'],
  logger: IntegrationLogger,
): true | false | undefined {
  if (CLOSED_DEVICE_STATUSES.includes(deviceStatus)) {
    return false;
  }
  if (OPEN_DEVICE_STATUSES.includes(deviceStatus)) {
    return true;
  }
  if (deviceStatus === undefined) {
    return undefined;
  }
  logger.warn({ deviceStatus }, 'Unexpected device status');
  return undefined;
}

export function calculateNumericSeverity(
  deviceStatus: DeviceConfigurationDeviceStatus['status'],
  logger: IntegrationLogger,
): number {
  if (!findingIsOpen(deviceStatus, logger)) {
    return 1;
  }
  switch (deviceStatus) {
    case 'nonCompliant':
    case 'error':
    case 'conflict':
      return 6;
    case 'unknown':
      return 4;
    default:
      logger.warn(
        { deviceStatus },
        'Unexpected device status when calculating numeric severity',
      );
      return 1;
  }
}

export function calculateSeverity(
  deviceStatus: DeviceConfigurationDeviceStatus['status'],
  logger: IntegrationLogger,
) {
  const numericSeverity = calculateNumericSeverity(deviceStatus, logger);
  if (numericSeverity === 0) {
    return 'informational';
  } else if (numericSeverity < 4) {
    return 'low';
  } else if (numericSeverity < 6) {
    return 'medium';
  } else if (numericSeverity < 8) {
    return 'high';
  } else if (numericSeverity <= 10) {
    return 'critical';
  } else {
    return 'unknown';
  }
}
