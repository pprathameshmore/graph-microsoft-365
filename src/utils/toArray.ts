import { isArray } from 'lodash';

/**
 * Convert a string or string array value to a string array
 */
export function toArray<T>(val: T | T[]): T[] {
  return isArray(val) ? (val as T[]) : [val as T];
}
