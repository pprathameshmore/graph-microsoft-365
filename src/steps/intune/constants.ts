import {
  generateRelationshipType,
  IntegrationError,
  RelationshipClass,
} from '@jupiterone/integration-sdk-core';
import { entities as activeDirectoryEntities } from '../active-directory';

export const steps: { [k: string]: string } = {
  FETCH_DEVICES: 'managed-devices',
  FETCH_COMPLIANCE_POLICIES_AND_FINDINGS: 'compliance-policies-and-findings',
  FETCH_DEVICE_CONFIGURATIONS_AND_FINDINGS:
    'device-configurations-and-findings',
  FETCH_MANAGED_APPLICATIONS: 'managed-applications',
  FETCH_DETECTED_APPLICATIONS: 'detected-applications',
};

export type ManagedDeviceType =
  | 'user_endpoint'
  | 'workstation'
  | 'laptop'
  | 'desktop'
  | 'computer'
  | 'server'
  | 'smartphone';
export const managedDeviceTypes: ManagedDeviceType[] = [
  'user_endpoint',
  'workstation',
  'laptop',
  'desktop',
  'computer',
  'server',
  'smartphone',
];
export const INTUNE_HOST_AGENT_KEY_PREFIX = 'intune-host-agent:';

export const entities = {
  MULTI_DEVICE: managedDeviceTypes.map((type) => {
    return {
      resourceName: 'Managed Device',
      _type: type,
      _class: ['Device', 'Host'], // Devices will not have a class of 'Device' if the device is not physical
    };
  }),
  HOST_AGENT: {
    resourceName: 'Intune Host Agent',
    _type: 'intune_host_agent',
    _class: 'HostAgent',
  },
  DEVICE_CONFIGURATION: {
    resourceName: 'Device Configuration',
    _type: 'intune_device_configuration',
    _class: ['Configuration', 'ControlPolicy'] as string[],
  },
  COMPLIANCE_POLICY: {
    resourceName: 'Compliance Policy',
    _type: 'intune_compliance_policy',
    _class: ['Configuration', 'ControlPolicy'] as string[],
  },
  NONCOMPLIANCE_FINDING: {
    resourceName: 'Noncompliance Finding',
    _type: 'intune_noncompliance_finding',
    _class: 'Finding',
  },
  MANAGED_APPLICATION: {
    resourceName: 'Managed Application',
    _type: 'intune_managed_application',
    _class: 'Application',
  },
  DETECTED_APPLICATION: {
    resourceName: 'Detected Application',
    _type: 'intune_detected_application',
    _class: 'Application',
  },
} as const;

function createRelationshipForAllDeviceTypes(relationshipMetadata: {
  sourceType?: string;
  targetType?: string;
  _class: RelationshipClass;
}) {
  const { sourceType, targetType, _class } = relationshipMetadata;
  if ((sourceType && targetType) || (!sourceType && !targetType)) {
    throw new IntegrationError({
      message: `'Invalid use of createRelationshipForAllDeviceTypes, you may have either a sourceType or a targetType'`,
      code: 'RELATIONSHIP_IMPLEMENTATION_ERROR',
    });
  }
  return managedDeviceTypes.map((type) => {
    return {
      _type: generateRelationshipType(
        _class,
        sourceType ?? type,
        targetType ?? type,
      ),
      _class,
      sourceType: sourceType ?? type,
      targetType: targetType ?? type,
    };
  });
}

export const relationships = {
  MULTI_USER_HAS_DEVICE: createRelationshipForAllDeviceTypes({
    sourceType: activeDirectoryEntities.USER._type,
    _class: RelationshipClass.HAS,
  }),
  MULTI_HOST_AGENT_MANAGES_DEVICE: createRelationshipForAllDeviceTypes({
    sourceType: entities.HOST_AGENT._type,
    _class: RelationshipClass.MANAGES,
  }),
  HOST_AGENT_ASSIGNED_DEVICE_CONFIGURATION: {
    _type: 'intune_host_agent_assigned_device_configuration',
    sourceType: entities.HOST_AGENT._type,
    _class: RelationshipClass.ASSIGNED,
    targetType: entities.DEVICE_CONFIGURATION._type,
  },
  HOST_AGENT_ASSIGNED_COMPLIANCE_POLICY: {
    _type: 'intune_host_agent_assigned_compliance_policy',
    sourceType: entities.HOST_AGENT._type,
    _class: RelationshipClass.ASSIGNED,
    targetType: entities.COMPLIANCE_POLICY._type,
  },
  DEVICE_CONFIGURATION_IDENTIFIED_NONCOMPLIANCE_FINDING: {
    _type: 'intune_device_configuration_identified_noncompliance_finding',
    sourceType: entities.DEVICE_CONFIGURATION._type,
    _class: RelationshipClass.IDENTIFIED,
    targetType: entities.NONCOMPLIANCE_FINDING._type,
  },
  COMPLIANCE_POLICY_IDENTIFIED_NONCOMPLIANCE_FINDING: {
    _type: 'intune_compliance_policy_identified_noncompliance_finding',
    sourceType: entities.COMPLIANCE_POLICY._type,
    _class: RelationshipClass.IDENTIFIED,
    targetType: entities.NONCOMPLIANCE_FINDING._type,
  },
  MULTI_DEVICE_HAS_NONCOMPLIANCE_FINDING: createRelationshipForAllDeviceTypes({
    targetType: entities.NONCOMPLIANCE_FINDING._type,
    _class: RelationshipClass.HAS,
  }),
  MULTI_DEVICE_ASSIGNED_MANAGED_APPLICATION: createRelationshipForAllDeviceTypes(
    {
      targetType: entities.MANAGED_APPLICATION._type,
      _class: RelationshipClass.ASSIGNED,
    },
  ),
  MULTI_DEVICE_INSTALLED_DETECTED_APPLICATION: createRelationshipForAllDeviceTypes(
    {
      targetType: entities.DETECTED_APPLICATION._type,
      _class: RelationshipClass.INSTALLED,
    },
  ),
  MANAGED_APPLICATION_MANAGES_DETECTED_APPLICATION: {
    _type: 'intune_managed_application_manages_detected_application',
    sourceType: entities.MANAGED_APPLICATION._type,
    _class: RelationshipClass.MANAGES,
    targetType: entities.DETECTED_APPLICATION._type,
  },
} as const;
