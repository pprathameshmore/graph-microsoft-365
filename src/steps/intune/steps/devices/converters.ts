import {
  createIntegrationEntity,
  createMappedRelationship,
  Entity,
  IntegrationError,
  Relationship,
  RelationshipDirection,
} from '@jupiterone/integration-sdk-core';
import {
  ComplianceState,
  DeviceType,
  ManagedDevice,
} from '@microsoft/microsoft-graph-types-beta';
import { entities as activeDirectoryEntities } from '../../../active-directory';
import { entities, INTUNE_HOST_AGENT_KEY_PREFIX } from '../../constants';
import { ManagedDeviceType, relationships } from '../../constants';

// https://docs.microsoft.com/en-us/graph/api/resources/intune-devices-manageddevice?view=graph-rest-1.0&viewFallbackFrom=graph-rest-beta
export function createManagedDeviceEntity(
  managedDevice: ManagedDevice,
): Entity {
  const isPhysicalDevice = isPhysical(managedDevice);
  const _class = ['Host'];
  if (isPhysicalDevice) {
    _class.push('Device');
  }
  return createIntegrationEntity({
    entityData: {
      source: managedDevice,
      assign: {
        _class,
        _type: selectDeviceType(managedDevice.deviceType, isPhysicalDevice),
        category: 'endpoint',
        id: managedDevice.id,
        name: managedDevice.deviceName,
        deviceName: managedDevice.deviceName,
        displayName: managedDevice.deviceName as string,
        deviceType: managedDevice.deviceType,
        model: managedDevice.model,
        make: managedDevice.manufacturer,
        version: managedDevice.model,
        serial: managedDevice.serialNumber,
        serialNumber: managedDevice.serialNumber,
        hardwareVendor:
          managedDevice.hardwareInformation?.manufacturer ??
          managedDevice.manufacturer,
        hardwareManufacturer:
          managedDevice.hardwareInformation?.manufacturer ??
          managedDevice.manufacturer,
        hardwareModel:
          managedDevice.hardwareInformation?.model ?? managedDevice.model,
        hardwareSerial:
          managedDevice.hardwareInformation?.serialNumber ??
          managedDevice.serialNumber,
        hardwareVersion:
          managedDevice.hardwareInformation?.model ?? managedDevice.model,
        meid: managedDevice.meid, // A mobile equipment identifier (MEID) is a globally unique number identifying a physical piece of CDMA2000 mobile station equipment
        imei: managedDevice.imei, // The International Mobile Equipment Identity (IMEI) is a number, usually unique, to identify 3GPP and iDEN mobile phones, as well as some satellite phones.
        iccid: managedDevice.iccid, // Every SIM card has a ICCID number, which stands for Integrated Circuit Card Identifier
        udid: managedDevice.udid, // UDID is an acronym for Unique Device Identifier.
        wifiMacAddress: managedDevice.wiFiMacAddress, // To communicate with a Wi-Fi network, a device must identify itself to the network using a unique network address called a Media Access Control (MAC) address
        easDeviceId: managedDevice.easActivated && managedDevice.easDeviceId, // Microsoft Exchange ActiveSync device identifier
        assetTag:
          managedDevice.azureActiveDirectoryDeviceId ??
          managedDevice.azureADDeviceId, // I believe these are legacy values now
        osDetails: `${managedDevice.operatingSystem} version ${managedDevice.osVersion}`,
        osName: managedDevice.operatingSystem,
        osVersion: managedDevice.osVersion,
        userEmails: managedDevice.emailAddress
          ? [managedDevice.emailAddress]
          : undefined,
        BYOD: managedDevice.managedDeviceOwnerType === 'personal',
        ownerType: managedDevice.managedDeviceOwnerType, // 'personal', 'company' or 'unknown'
        encrypted: managedDevice.isEncrypted,
        userId: managedDevice.userId,
        userDisplayName: managedDevice.userDisplayName,
        phoneNumber: managedDevice.phoneNumber,
        managed:
          managedDevice.managementState &&
          managedDevice.managementState !== 'discovered',
        supervised: managedDevice.isSupervised,
        jailBroken: managedDevice.jailBroken !== 'False',
        username: managedDevice.userPrincipalName,
        physical: isPhysicalDevice,
        // POTENTIAL: managedDevice.usersLoggedOn - link out to other users perhaps?
      },
    },
  });
}

export function createIntuneHostAgentEntity(
  managedDevice: ManagedDevice,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: managedDevice,
      assign: {
        _class: entities.HOST_AGENT._class,
        function: ['endpoint-compliance', 'endpoint-protection'], // https://github.com/JupiterOne/data-model/blob/master/src/schemas/HostAgent.json
        _type: entities.HOST_AGENT._type,
        _key: INTUNE_HOST_AGENT_KEY_PREFIX + managedDevice.id,
        name: 'intune-host-agent',
        displayName: 'Intune Host Agent',
        managementAgent: managedDevice.managementAgent, // Management channel of the device. Possible values include: eas, mdm, easMdm, intuneClient, easIntuneClient, jamf, googleCloudDevicePolicyController, ...
        state: managedDevice.managementState,
        registrationState: managedDevice.deviceRegistrationState,
        complianceState: managedDevice.complianceState,
        compliant: isCompliant(managedDevice.complianceState),
      },
    },
  });
}

/**
 * Converts Microsoft's compliance state to JupiterOne's `compliant` boolean.
 * There are some complianceStates where we can not determine if the device is compliant or not.
 * In those cases this key should remain `undefined`.
 */
function isCompliant(complianceState?: ComplianceState) {
  return ['unknown', 'configManager', undefined].includes(complianceState)
    ? undefined
    : complianceState === 'compliant';
}

/**
 * Is the device a physical device or a virtual server
 */
function isPhysical(managedDevice: ManagedDevice) {
  return managedDevice.model
    ? managedDevice.model !== 'Virtual Machine'
    : undefined; // https://docs.microsoft.com/en-us/mem/intune/fundamentals/windows-10-virtual-machines#reporting
}

/**
 * Select the correct device _type based on the Microsoft DeviceType
 * types determined from managed-questions-endpoint.yaml
 */
export function selectDeviceType(
  deviceType?: DeviceType,
  isPhysical?: Boolean,
): ManagedDeviceType {
  if (!isPhysical) {
    return 'server';
  }
  switch (deviceType) {
    case 'desktop':
    case 'mac':
    case 'unix':
    case 'linux':
      return 'desktop';
    case 'windowsRT': // retired mobile OS
    case 'winMO6': // retired mobile OS
    case 'nokia':
    case 'iPhone':
    case 'iPod':
    case 'android':
    case 'androidForWork': // same OS as android, idk why it is differentiated here
    case 'androidEnterprise': // same OS as android, idk why it is differentiated here
    case 'androidnGMS': // same OS as android, idk why it is differentiated here
    case 'blackberry':
    case 'palm':
      return 'smartphone';
    case 'winCE':
    case 'winEmbedded':
      return 'workstation';
    case 'macMDM':
    case 'windows10x':
      return 'laptop';
    case 'cloudPC':
      return 'server';
    case 'iSocConsumer': // some open source thing
    case 'holoLens': // augmented reality tool
    case 'surfaceHub': // smart whiteboard
    case 'unknown':
      return 'user_endpoint';
    default:
      return 'user_endpoint';
  }
}

/**
 * Creates a mapped relationship between a User and a Device if userId or email is provided.
 * Should be used when the active directory is not being ingested in this integration instance.
 */
export function createUserDeviceMappedRelationship(
  managedDeviceEntity: Entity,
  userId?: string,
  email?: string,
): Relationship | undefined {
  const relationship = relationships.MULTI_USER_HAS_DEVICE.find(
    (relationship) =>
      [relationship.sourceType, relationship.targetType].includes(
        managedDeviceEntity._type,
      ),
  );
  if (!relationship) {
    throw new IntegrationError({
      message: `Could not find relationship for deviceType ${managedDeviceEntity._type}`,
      code: 'RELATIONSHIP_IMPLEMENTATION_ERROR',
    });
  }
  return userId || email
    ? createMappedRelationship({
        _class: relationship._class,
        _type: relationship._type,
        _mapping: {
          sourceEntityKey: managedDeviceEntity._key,
          relationshipDirection: RelationshipDirection.REVERSE,
          targetFilterKeys: [
            ['_class', 'id'],
            ['_class', 'email'],
          ],
          targetEntity: {
            _class: activeDirectoryEntities.USER._class,
            _type: activeDirectoryEntities.USER._type,
            _key: userId,
            id: userId,
            email: email,
          },
        },
      })
    : undefined;
}
