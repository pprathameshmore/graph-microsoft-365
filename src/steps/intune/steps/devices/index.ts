import {
  createDirectRelationship,
  IntegrationStepExecutionContext,
  Step,
} from '@jupiterone/integration-sdk-core';
import { IntegrationConfig, IntegrationStepContext } from '../../../../types';
import { steps as activeDirectorySteps } from '../../../active-directory';
import { DeviceManagementIntuneClient } from '../../clients/deviceManagementIntuneClient';
import { relationships, entities, steps } from '../../constants';
import {
  createManagedDeviceEntity,
  createIntuneHostAgentEntity,
  createUserDeviceMappedRelationship,
} from './converters';

/**
 * Intune ManagedDevices are componsed of a physical or virtual device plus the Intune host agent.
 * This agent allows users to apply configurations and policies to their managed devices. These
 * need to be modeled as seperate entities as devices may be shared between multiple integrations
 * where the Intune host agent is unique to this integration.
 */
export async function fetchDevices(
  executionContext: IntegrationStepContext,
): Promise<void> {
  const { logger, instance, jobState } = executionContext;
  const intuneClient = new DeviceManagementIntuneClient(
    logger,
    instance.config,
  );

  await intuneClient.iterateManagedDevices(async (device) => {
    const deviceEntity = createManagedDeviceEntity(device);
    await jobState.addEntity(deviceEntity);
    const userEntity = await jobState.findEntity(device.userId as string);
    const userDeviceRelationship = userEntity
      ? createDirectRelationship({
          _class: relationships.MULTI_USER_HAS_DEVICE[0]._class,
          from: userEntity,
          to: deviceEntity,
        })
      : createUserDeviceMappedRelationship(
          deviceEntity,
          device.userId as string,
          device.emailAddress as string,
        );
    if (userDeviceRelationship) {
      await jobState.addRelationship(userDeviceRelationship);
    }

    // Create and relate Intune Host Agent based on Managed Device
    const hostAgentEntity = createIntuneHostAgentEntity(device);
    await jobState.addEntity(hostAgentEntity);
    const deviceHostAgentRelationship = createDirectRelationship({
      _class: relationships.MULTI_HOST_AGENT_MANAGES_DEVICE[0]._class,
      from: hostAgentEntity,
      to: deviceEntity,
    });
    await jobState.addRelationship(deviceHostAgentRelationship);
  });
}

export const deviceSteps: Step<
  IntegrationStepExecutionContext<IntegrationConfig>
>[] = [
  {
    id: steps.FETCH_DEVICES,
    name: 'Managed Devices',
    entities: [...entities.MULTI_DEVICE, entities.HOST_AGENT],
    relationships: [
      ...relationships.MULTI_USER_HAS_DEVICE,
      ...relationships.MULTI_HOST_AGENT_MANAGES_DEVICE,
    ],
    dependsOn: [activeDirectorySteps.FETCH_USERS],
    executionHandler: fetchDevices,
  },
];
