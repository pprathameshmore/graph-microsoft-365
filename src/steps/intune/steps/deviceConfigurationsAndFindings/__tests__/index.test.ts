import {
  createMockStepExecutionContext,
  Recording,
} from '@jupiterone/integration-sdk-testing';
import { setupAzureRecording } from '../../../../../../test/recording';
import { config } from '../../../../../../test/config';
import { fetchDeviceConfigurationsAndFindings } from '..';
import { fetchDevices } from '../../devices';
import { entities, relationships } from '../../../constants';
import { Entity } from '@jupiterone/integration-sdk-core';
import { isEqual } from 'lodash';
import { toArray } from '../../../../../utils/toArray';

let recording: Recording;

afterEach(async () => {
  if (recording) {
    await recording.stop();
  }
});

describe('fetchDeviceConfigurationsAndFindings', () => {
  test('should make entities and relationships correctly', async () => {
    recording = setupAzureRecording({
      directory: __dirname,
      name: 'fetchDeviceConfigurationsAndFindings',
    });
    const context = createMockStepExecutionContext({ instanceConfig: config });

    await fetchDevices(context);
    await fetchDeviceConfigurationsAndFindings(context);

    const deviceConfigEntities = context.jobState.collectedEntities.filter(
      (e) => isEqual(e._class, toArray(entities.DEVICE_CONFIGURATION._class)),
    );
    const noncomplianceFindingEntities = context.jobState.collectedEntities.filter(
      (e) => isEqual(e._class, toArray(entities.NONCOMPLIANCE_FINDING._class)),
    );
    const hostAgentDeviceConfigRelationships = context.jobState.collectedRelationships.filter(
      (r) =>
        isEqual(
          r._type,
          relationships.HOST_AGENT_ASSIGNED_DEVICE_CONFIGURATION._type,
        ),
    );
    const deviceConfigFindingRelationships = context.jobState.collectedRelationships.filter(
      (r) =>
        isEqual(
          r._type,
          relationships.DEVICE_CONFIGURATION_IDENTIFIED_NONCOMPLIANCE_FINDING
            ._type,
        ),
    );
    const deviceFindingRelationships = context.jobState.collectedRelationships.filter(
      (r) =>
        relationships.MULTI_DEVICE_HAS_NONCOMPLIANCE_FINDING.map(
          (c) => c._type,
        ).includes(r._type),
    );

    // Check that we have Device Configurations
    expect(deviceConfigEntities.length).toBeGreaterThan(0);
    expect(deviceConfigEntities).toMatchGraphObjectSchema({
      _class: entities.DEVICE_CONFIGURATION._class,
    });
    expect(deviceConfigEntities).toMatchSnapshot('deviceConfigurationEntities');
    // Check that there are no orphaned Device Configuraitons
    deviceConfigEntities.forEach((configEntity) => {
      expect(
        hostAgentDeviceConfigRelationships.find((r) =>
          r._key.includes(configEntity._key),
        ),
      ).toBeTruthy();
    });

    // Check that we have Noncompliance Findings
    expect(noncomplianceFindingEntities.length).toBeGreaterThan(0);
    expect(noncomplianceFindingEntities).toMatchGraphObjectSchema({
      _class: entities.NONCOMPLIANCE_FINDING._class,
    });
    expect(noncomplianceFindingEntities).toMatchSnapshot(
      'noncomplianceFindingEntities',
    );
    // Check that only open findings are created
    noncomplianceFindingEntities.forEach((entity: Entity) => {
      expect(entity.open).toBe(true);
    });

    // Check that we have DEVICE_USES_DEVICE_CONFIGURATION relationships
    expect(hostAgentDeviceConfigRelationships.length).toBeGreaterThan(0);
    expect(hostAgentDeviceConfigRelationships).toMatchDirectRelationshipSchema({
      schema: {
        properties: {
          complianceStatus: { type: 'string' },
          compliant: { type: 'boolean' },
        },
      },
    });
    expect(hostAgentDeviceConfigRelationships).toMatchSnapshot(
      'noncomplianceFindingDeviceRelationships',
    );

    // Check that we have DEVICE_CONFIGURATION_IDENTIFIED_NONCOMPLIANCE_FINDING relationships
    expect(deviceConfigFindingRelationships.length).toBeGreaterThan(0);
    expect(deviceConfigFindingRelationships).toMatchDirectRelationshipSchema(
      {},
    );
    expect(deviceConfigFindingRelationships).toMatchSnapshot(
      'noncomplianceFindingDeviceConfigurationRelationships',
    );

    // Check that we have DEVICE_HAS_NONCOMPLIANCE_FINDING relationships
    expect(deviceFindingRelationships.length).toBeGreaterThan(0);
    expect(deviceFindingRelationships).toMatchDirectRelationshipSchema({});
    expect(deviceFindingRelationships).toMatchSnapshot(
      'deviceDeviceConfigurationRelationships',
    );
  });
});
