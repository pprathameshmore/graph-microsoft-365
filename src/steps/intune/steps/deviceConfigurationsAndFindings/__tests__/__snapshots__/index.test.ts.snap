// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchDeviceConfigurationsAndFindings should make entities and relationships correctly: deviceConfigurationEntities 1`] = `
Array [
  Object {
    "_class": Array [
      "Configuration",
      "ControlPolicy",
    ],
    "_key": "6aa605f0-744e-4ae1-8031-c6564624c89b",
    "_rawData": Array [
      Object {
        "name": "default",
        "rawData": Object {
          "@odata.type": "#microsoft.graph.iosCustomConfiguration",
          "createdDateTime": "2021-02-17T18:31:21.3225031Z",
          "description": null,
          "displayName": "Test configuration policy",
          "id": "6aa605f0-744e-4ae1-8031-c6564624c89b",
          "lastModifiedDateTime": "2021-02-17T18:31:21.3225031Z",
          "payload": "MTk3",
          "payloadFileName": "lastSuccessfulBuildId.txt",
          "payloadName": "Test fake profile",
          "version": 1,
        },
      },
    ],
    "_type": "intune_device_configuration",
    "category": "config",
    "configurationType": "iosCustomConfiguration",
    "createdOn": 1613586681322,
    "description": "Test configuration policy",
    "displayName": "Test configuration policy",
    "function": "endpoint-configuration",
    "id": "6aa605f0-744e-4ae1-8031-c6564624c89b",
    "name": "Test configuration policy",
    "version": 1,
  },
]
`;

exports[`fetchDeviceConfigurationsAndFindings should make entities and relationships correctly: deviceDeviceConfigurationRelationships 1`] = `
Array [
  Object {
    "_class": "HAS",
    "_fromEntityKey": "683dbff1-c8ff-4996-91ae-85484de46cfc",
    "_key": "683dbff1-c8ff-4996-91ae-85484de46cfc|has|7b521133-f056-40a9-b639-b76afd67e4bb_6aa605f0-744e-4ae1-8031-c6564624c89b_683dbff1-c8ff-4996-91ae-85484de46cfc",
    "_toEntityKey": "7b521133-f056-40a9-b639-b76afd67e4bb_6aa605f0-744e-4ae1-8031-c6564624c89b_683dbff1-c8ff-4996-91ae-85484de46cfc",
    "_type": "smartphone_has_intune_noncompliance_finding",
    "displayName": "HAS",
  },
]
`;

exports[`fetchDeviceConfigurationsAndFindings should make entities and relationships correctly: noncomplianceFindingDeviceConfigurationRelationships 1`] = `
Array [
  Object {
    "_class": "IDENTIFIED",
    "_fromEntityKey": "6aa605f0-744e-4ae1-8031-c6564624c89b",
    "_key": "6aa605f0-744e-4ae1-8031-c6564624c89b|identified|7b521133-f056-40a9-b639-b76afd67e4bb_6aa605f0-744e-4ae1-8031-c6564624c89b_683dbff1-c8ff-4996-91ae-85484de46cfc",
    "_toEntityKey": "7b521133-f056-40a9-b639-b76afd67e4bb_6aa605f0-744e-4ae1-8031-c6564624c89b_683dbff1-c8ff-4996-91ae-85484de46cfc",
    "_type": "intune_device_configuration_identified_noncompliance_finding",
    "displayName": "IDENTIFIED",
  },
]
`;

exports[`fetchDeviceConfigurationsAndFindings should make entities and relationships correctly: noncomplianceFindingDeviceRelationships 1`] = `
Array [
  Object {
    "_class": "ASSIGNED",
    "_fromEntityKey": "intune-host-agent:683dbff1-c8ff-4996-91ae-85484de46cfc",
    "_key": "intune-host-agent:683dbff1-c8ff-4996-91ae-85484de46cfc|assigned|6aa605f0-744e-4ae1-8031-c6564624c89b",
    "_toEntityKey": "6aa605f0-744e-4ae1-8031-c6564624c89b",
    "_type": "intune_host_agent_assigned_device_configuration",
    "complianceStatus": "error",
    "compliant": false,
    "displayName": "ASSIGNED",
  },
]
`;

exports[`fetchDeviceConfigurationsAndFindings should make entities and relationships correctly: noncomplianceFindingEntities 1`] = `
Array [
  Object {
    "_class": Array [
      "Finding",
    ],
    "_key": "7b521133-f056-40a9-b639-b76afd67e4bb_6aa605f0-744e-4ae1-8031-c6564624c89b_683dbff1-c8ff-4996-91ae-85484de46cfc",
    "_rawData": Array [
      Object {
        "name": "default",
        "rawData": Object {
          "complianceGracePeriodExpirationDateTime": "2021-02-17T19:13:41.8113Z",
          "deviceDisplayName": "[REDACTED]",
          "deviceModel": null,
          "id": "7b521133-f056-40a9-b639-b76afd67e4bb_6aa605f0-744e-4ae1-8031-c6564624c89b_683dbff1-c8ff-4996-91ae-85484de46cfc",
          "lastReportedDateTime": "2021-03-02T16:23:45.8206923Z",
          "status": "error",
          "userName": "[REDACTED]",
          "userPrincipalName": "[REDACTED]",
        },
      },
    ],
    "_type": "intune_noncompliance_finding",
    "active": undefined,
    "assessment": "Test configuration policy",
    "category": "endpoint",
    "createdOn": undefined,
    "displayName": "Latest finding from Test configuration policy",
    "id": "7b521133-f056-40a9-b639-b76afd67e4bb_6aa605f0-744e-4ae1-8031-c6564624c89b_683dbff1-c8ff-4996-91ae-85484de46cfc",
    "lastProcessedOn": 1614702225820,
    "lastTestedOn": 1614702225820,
    "lastUpdatedOn": 1614702225820,
    "name": "Latest finding from Test configuration policy",
    "numericSeverity": 6,
    "open": true,
    "severity": "high",
    "status": "error",
  },
]
`;
